src/graphql/graphql-api.php

```diff
@@ -0,0 +1,245 @@
+<?php
+/**
+ * GraphQL API
+ *
+ * @package vip-block-data-api
+ */
+
+namespace WPCOMVIP\BlockDataApi;
+
+defined( 'ABSPATH' ) || die();
+
+/**
+ * GraphQL API to offer an alternative to the REST API.
+ */
+class GraphQLApi {
+	/**
+	 * Initiatilize the graphQL API, if its allowed
+	 *
+	 * @access private
+	 */
+	public static function init() {
+		/**
+		 * Filter to enable/disable the graphQL API. By default, it is enabled.
+		 * 
+		 * @param bool $is_graphql_to_be_enabled Whether the graphQL API should be enabled or not.
+		 */
+		$is_graphql_to_be_enabled = apply_filters( 'vip_block_data_api__is_graphql_enabled', true );
+
+		if ( ! $is_graphql_to_be_enabled ) {
+			return;
+		}
+
+		add_filter( 'vip_block_data_api__sourced_block_result', [ __CLASS__, 'transform_block_format' ], 10, 5 );
+
+		add_action( 'graphql_register_types', [ __CLASS__, 'register_types' ] );
+	}
+
+	/**
+	 * Extract the blocks data for a post, and return back in the format expected by the graphQL API.
+	 *
+	 * @param  WPGraphQL\Model\Post $post_model Post model for post.
+	 * @return array
+	 */
+	public static function get_blocks_data( $post_model ) {
+		$post_id        = $post_model->ID;
+		$post           = get_post( $post_id );
+		$filter_options = [ 'graphQL' => true ];
+
+		$content_parser = new ContentParser();
+
+		$parser_results = $content_parser->parse( $post->post_content, $post_id, $filter_options );
+
+		// We need to not return a WP_Error object, and so a regular exception is returned.
+		if ( is_wp_error( $parser_results ) ) {
+			Analytics::record_error( $parser_results );
+
+			// Return API-safe error with extra data (e.g. stack trace) removed.
+			return new \Exception( $parser_results->get_error_message() );
+		}
+
+		return $parser_results;
+	}
+
+	/**
+	 * Transform the block's format to the format expected by the graphQL API.
+	 *
+	 * @param array  $sourced_block An associative array of parsed block data with keys 'name' and 'attribute'.
+	 * @param string $block_name Name of the parsed block, e.g. 'core/paragraph'.
+	 * @param int    $post_id Post ID associated with the parsed block.
+	 * @param array  $block Result of parse_blocks() for this block. Contains 'blockName', 'attrs', 'innerHTML', and 'innerBlocks' keys.
+	 * @param array  $filter_options Options to filter using, if any.
+	 *
+	 * @return array
+	 */
+	public static function transform_block_format( $sourced_block, $block_name, $post_id, $block, $filter_options ) { // phpcs:disable Generic.CodeAnalysis.UnusedFunctionParameter.FoundAfterLastUsed
```

Comment:
I don't understand why this is called via a filter from `ContentParser`. Why not just call it directly in `get_blocks_data`?

Reply:
There's 2 reasons why it was done that way:

- I wanted to keep the GraphQL logic separate from the parsing logic
- I also wanted to ensure that, the transforming happened at the same time as when the parser worked so we didn't loop over each block twice

So this way, while its odd, gives us the best of both worlds. With that said, I do think it can be done in a better way even if that means we go over the entire block list twice.

Talking it over with @alecgeatches, I'll change it up to:

- Get the parser results from ContentParser
- Go over the results, and transform it into the graphQL format
- This would make any potential changes to the structure, like flattening or not flattening innerBlocks easier as well as anything else we think of.
- No more filter oddeties to be done

Reply:
Switched it over so the transformation is done after the parser comes back, so more filter weirdness